pandas/core/groupby.py

```diff
@@ -3465,19 +3459,23 @@ def transform(self, func, *args, **kwargs):
         if not result.columns.equals(obj.columns):
             return self._transform_general(func, *args, **kwargs)
 
-        results = np.empty_like(obj.values, result.values.dtype)
-        for (name, group), (i, row) in zip(self, result.iterrows()):
-            indexer = self._get_index(name)
-            if len(indexer) > 0:
-                results[indexer] = np.tile(row.values, len(
-                    indexer)).reshape(len(indexer), -1)
+        # Fast transform path for aggregations
 
+        # if there were groups with no observations (Categorical only?)
+        # try casting data to original dtype
         counts = self.size().fillna(0).values
-        if any(counts == 0):
-            results = self._try_cast(results, obj[result.columns])
+        cast = (counts == 0).any()
+
+        # by column (could be by block?) reshape aggregated data to
+        # size of original frame by repeating obvservations with take
+        ids, _, ngroup = self.grouper.group_info
+        out = {}
```

Comment:
this is going to fail with dupe columns, better to just append the arrays to a list.